cmd/promoted-image-governor/main.go

```diff
@@ -0,0 +1,255 @@
+}
+
+func loadOpenshiftMappingConfig(path string) (*OpenshiftMappingConfig, error) {
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read file: %s", path)
+	}
+	openshiftMappingConfig := &OpenshiftMappingConfig{}
+	if err = yaml.Unmarshal(data, openshiftMappingConfig); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal the config %q: %w", string(data), err)
+	}
+	return openshiftMappingConfig, nil
+}
+
+func newOpts() *options {
+	opts := &options{}
+	fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+	// Controller-Runtimes root package imports the package that sets this flag
+	kubeconfigFlagDescription := "Path to the kubeconfig file to use for CLI requests."
+	if f := fs.Lookup("kubeconfig"); f != nil {
+		f.Usage = kubeconfigFlagDescription
+		// https://i.kym-cdn.com/entries/icons/original/000/018/012/this_is_fine.jpeg
+		defer func() { opts.kubeconfig = f.Value.String() }()
+	} else {
+		fs.StringVar(&opts.kubeconfig, "kubeconfig", "", kubeconfigFlagDescription)
+	}
+	fs.StringVar(&opts.ciOperatorconfigPath, "ci-operator-config-path", "", "Path to the ci operator config")
+	fs.BoolVar(&opts.dryRun, "dry-run", true, "Whether to run the controller-manager with dry-run")
+	fs.Var(&opts.ignoredImageStreamTagsRaw, "ignored-image-stream-tags", "An imagestreamtag to ignore while checking. It must be in namespace/name:tag format (e.G `ci/clonerefs:latest`). Can be passed multiple times.")
+	fs.StringVar(&opts.openshiftMappingDir, "openshift-mapping-dir", "", "Path to the openshift mapping directory")
+	fs.StringVar(&opts.openshiftMappingConfigPath, "openshift-mapping-config", "", "Path to the openshift mapping config file")
+	if err := fs.Parse(os.Args[1:]); err != nil {
+		logrus.WithError(err).Fatal("could not parse input")
+	}
+	opts.ignoredImageStreamTags = opts.ignoredImageStreamTagsRaw.StringSet()
+	if opts.openshiftMappingConfigPath != "" {
+		c, err := loadOpenshiftMappingConfig(opts.openshiftMappingConfigPath)
+		if err != nil {
+			logrus.WithError(err).Fatal("could not load openshift mapping config")
+		}
+		opts.openshiftMappingConfig = c
+	}
+	return opts
+}
+
+func (o *options) validate() error {
+	if o.ciOperatorconfigPath == "" {
+		return fmt.Errorf("--ci-operator-config-path must be set")
+	}
+	if (o.openshiftMappingDir == "") != (o.openshiftMappingConfigPath == "") {
+		return fmt.Errorf("--openshift-mapping-dir and --openshift-mapping-config must be set together")
+	}
+	return nil
+}
+
+func addSchemes() error {
+	if err := imagev1.AddToScheme(scheme.Scheme); err != nil {
+		return fmt.Errorf("failed to add imagev1 to scheme: %w", err)
+	}
+	return nil
+}
+
+func validatePromotedTags(tags []api.ImageStreamTagReference) error {
+	tagCount := map[api.ImageStreamTagReference]int{}
+	for _, tag := range tags {
+		if _, ok := tagCount[tag]; ok {
+			tagCount[tag] = tagCount[tag] + 1
+		} else {
+			tagCount[tag] = 1
+		}
+	}
+	dups := sets.NewString()
```

Comment:
Tags are keys in a map - they must be unique. No need for a set.

Reply:
Updated.